from Bio import SeqIO
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.Alphabet import IUPAC
from Bio.SeqFeature import SeqFeature, FeatureLocation
 
# Relevant feature information for a prokaryotic genomic sequence encoding a protein:
# coding region intervals, including start and stop codons, if present
# protein name
# gene name, if known
# amino acid sequence, if known

# Create a sequence
sequence_string = "ggggaaaattttaaaaccccaaaa"
sequence_object = Seq(sequence_string, IUPAC.unambiguous_dna)
 
# Create a record
record = SeqRecord(sequence_object,
                   id='123456789', # random accession number
                   name='Example',
                   description='An example GenBank file generated by BioPython')
 
# Add annotation
feature = SeqFeature(location=FeatureLocation(start=3, end=12),
                     type='gene',
                     qualifiers={'product': 'hello'})
record.features.append(feature)
 
# Save as GenBank file
output_file = open('example.gb', 'w')
SeqIO.write(record, output_file, 'genbank')

# output =  "LOCUS       Example\t24bp\tDNA\tPHG\tcurrent date\n"
# output += "DEFINITION  An example description\n"
# output += "ACCESSION   12345678\n"
# output += "VERSION     12345678\n"
# output += "SOURCE      .\n"
# output += "  ORGANISM  ."

# with open('example1.gb', 'w') as outfile:
#     outfile.write(output)